### PR TITLE
pulp_server Debian repo support

### PR DESCRIPTION
**Pull request by johnwarburton**
##### SUMMARY

Add Debian repo support for pulp_server

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

pulp_server

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: plugins/modules/packaging/os/pulp_repo.py [hunk 1]
```diff
@@ -22,6 +22,18 @@
       - Whether or not to add the export distributor to new C(rpm) repositories.
     type: bool
     default: no
+  components:
+    description:
+      - Debian repo components.
+    type: str
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pulp_repo.py:28]:
```suggestion
    type: str
    version_added: 2.0.0
```
This needs to be added for all new options.

File: plugins/modules/packaging/os/pulp_repo.py [hunk 2]
```diff
@@ -117,7 +129,7 @@
     type: str
   repo_type:
     description:
-      - Repo plugin type to use (i.e. C(rpm), C(docker)).
+      - Repo plugin type to use (i.e. C(rpm), C(deb)).
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pulp_repo.py:132]: Is `docker` not a valid value? Or why did you remove it?

File: plugins/modules/packaging/os/pulp_repo.py [hunk 3]
```diff
@@ -194,6 +206,21 @@
     name: my_old_repo
     repo_type: rpm
     state: absent
+
+- name: Copy of local internal repo with newer version of ansible for bionic, with legacy http support
+  community.general.pulp_repo:
+    name: ansible-bionic
+    description: 'Ubuntu 18 repo for anisble 2.9.15'
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pulp_repo.py:213]:
```suggestion
    description: 'Ubuntu 18.04 repo for Ansible 2.9.15'
```

File: plugins/modules/packaging/os/pulp_repo.py [hunk 4]
```diff
@@ -194,6 +206,21 @@
     name: my_old_repo
     repo_type: rpm
     state: absent
+
+- name: Copy of local internal repo with newer version of ansible for bionic, with legacy http support
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pulp_repo.py:210]:
```suggestion
- name: Copy of local internal repo with newer version of Ansible for bionic, with legacy HTTP support
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! Here are some first comments.

You need to add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to).

File: plugins/modules/packaging/os/pulp_repo.py (hunk 2 above)
Comment (johnwarburton) [reply plugins/modules/packaging/os/pulp_repo.py:132]: Hi. It was never a valid value. Only 'rpm' was ever supported. Having it there made reading this confusing as to what was supported

File: plugins/modules/packaging/os/pulp_repo.py (hunk 3 above)
Comment (johnwarburton) [reply plugins/modules/packaging/os/pulp_repo.py:213]: Done

File: plugins/modules/packaging/os/pulp_repo.py (hunk 4 above)
Comment (johnwarburton) [reply plugins/modules/packaging/os/pulp_repo.py:210]: Done

File: changelogs/fragments/1408-pulp_server-debian-repo-support.yml [hunk 5]
```diff
@@ -0,0 +1,3 @@
+minor_changes:
+- support for 'deb' repo type in pulp_repo.py
+- added description parameter for repo
```
Comment (felixfontein) [inline changelogs/fragments/1408-pulp_server-debian-repo-support.yml:3]:
```suggestion
- pulp_repo - support for ``deb`` repo type; added options ``components``, ``releases`` and ``description`` (https://github.com/ansible-collections/community.general/pull/1408).
```

File: plugins/modules/packaging/os/pulp_repo.py [hunk 6]
```diff
@@ -341,6 +376,65 @@ def create_repo(
...
             }
 
+        if self.repo_type == 'deb':
+            deb_distributor = dict()
+            deb_distributor['distributor_id'] = "deb_distributor"
+            deb_distributor['distributor_type_id'] = "deb_distributor"
+            deb_distributor['auto_publish'] = True
+            deb_distributor['distributor_config'] = dict()
+            deb_distributor['distributor_config']['http'] = serve_http
+            deb_distributor['distributor_config']['https'] = serve_https
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pulp_repo.py:386]: There seems to be **a lot** of shared code between the code paths. I think it would make sense to consolidate some of it.

File: plugins/modules/packaging/os/pulp_repo.py [hunk 7]
```diff
@@ -688,6 +794,9 @@ def main():
             # we set the whole config at the same time.
             if not server.compare_repo_importer_config(
                 repo,
+                components=components,
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pulp_repo.py:797]: Did you check whether idempotency / updating actually works? Both for `rpm` and `deb` repositories?

**Comment by felixfontein** [comment]
@johnwarburton if you have questions on the review comments, please feel free to ask. If you need help with something, also please ask.

needs_info

**Comment by felixfontein** [comment]
@johnwarburton do you still want to finish this?

needs_info

**Comment by felixfontein** [comment]
Closing due to inactivity.

**Comment by felixfontein** [comment]
Please ping us if you want to continue this PR in the future. If anyone else is interested in working on this, please create a new PR based on the current state of this PR and mention it here.
